### PR TITLE
NIFI-6255 NIFI-6287: Hash function for expression language and record…

### PR DESCRIPTION
**Pull request by pcgrenier**
… path.

Expression function = ${field.value:hash('MD5')}
Record Path function = hash(/field, 'MD5')

#### Description of PR

hash functions

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [x] Is there a JIRA ticket associated with this PR? Is it referenced 
     in the commit message?

- [x] Does your PR title start with **NIFI-XXXX** where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.

- [x] Has your PR been rebased against the latest commit within the target branch (typically `master`)?

- [x] Is your initial contribution a single, squashed commit? _Additional commits in response to PR reviewer feedback should be made on this branch and pushed to allow change tracking. Do not `squash` or use `--force` when pushing to allow for clean monitoring of changes._

### For code changes:
- [x] Have you ensured that the full suite of tests is executed via `mvn -Pcontrib-check clean install` at the root `nifi` folder?
- [x] Have you written or updated unit tests to verify your changes?
- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)? 
- [ ] If applicable, have you updated the `LICENSE` file, including the main `LICENSE` file under `nifi-assembly`?
- [ ] If applicable, have you updated the `NOTICE` file, including the main `NOTICE` file found under `nifi-assembly`?
- [ ] If adding new Properties, have you added `.displayName` in addition to .name (programmatic access) for each of the new properties?

### For documentation related changes:
- [x] Have you ensured that format looks appropriate for the output in which it is rendered?

### Note:
Please ensure that once the PR is submitted, you check travis-ci for build issues and submit an update to your PR as soon as possible.

### CODE REVIEW

File: nifi-commons/nifi-expression-language/src/test/java/org/apache/nifi/attribute/expression/language/TestQuery.java [hunk 1]
```diff
@@ -1816,6 +1816,18 @@ public void testIfElse() {
...
+    @Test
+    public void testHash() {
+        final Map<String, String> attributes = new HashMap<>();
+        attributes.put("str_attr", "string value");
+        attributes.put("nbr_attr", "10");
+        verifyEquals("${literal('john'):hash('MD5')}", attributes, "527bd5b5d689e2c32ae974c6229ff785");
+        verifyEquals("${str_attr:hash('MD5')}", attributes, "64e58419496c7248b4ef25731f88b8c3");
+        verifyEquals("${str_attr:hash('SHA-1')}", attributes, "34990db823e7bb2b47278a7fbf08c62d9e8e4307");
+        verifyEquals("${str_attr:hash('SHA-256')}", attributes, "9b6a1a9167a5caf3f5948413faa89e0ec0de89e12bef55327442e60dcc0e8c9b");
+        verifyEquals("${nbr_attr:toNumber():hash('MD5')}", attributes, "d3d9446802a44259755d38e6d163e820");
```
Comment (alopresto) [inline nifi-commons/nifi-expression-language/src/test/java/org/apache/nifi/attribute/expression/language/TestQuery.java]: I would add a line verifying that `${nbr_attr:hash('MD5')}` also equals the same hash as when it is converted using `toNumber()` to demonstrate the equivalence.

**Comment by alopresto** [comment]
Thanks for contributing this functionality. My only question is what happens when the user provides an unsupported algorithm name or doesn't format it explicitly as supported by Java's `MessageDigest`? Can we improve the messaging in the errors and provide an enumerated list of the exact strings to use in the documentation for both Expression Language and RecordPath syntax?

**Comment by pcgrenier** [comment]
> Thanks for contributing this functionality. My only question is what happens when the user provides an unsupported algorithm name or doesn't format it explicitly as supported by Java's `MessageDigest`? Can we improve the messaging in the errors and provide an enumerated list of the exact strings to use in the documentation for both Expression Language and RecordPath syntax?

as is it returns an error, saying unsupported algorithm name: the_name given.

Does the list change? [message digest docs](https://docs.oracle.com/javase/7/docs/technotes/guides/security/StandardNames.html#MessageDigest). Because I would have to update the list to match the java version. The alternative is a utility class to pull the security providers out and list them in the error message.

**Comment by alopresto** [comment]
I don't think the error has to explicitly list all the available options, but I think the docs should. The error can just reference the docs to let people know where the definitive list is. 

The list of supported algorithms does not change frequently, and the docs can be updated in the event the list changes. 

There are some examples of our current hash usage in [`HashService`](https://github.com/apache/nifi/blob/master/nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/security/util/crypto/HashService.java#L59) and [`HashAlgorithm`](https://github.com/apache/nifi/blob/master/nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/security/util/crypto/HashAlgorithm.java). I believe in the docs for `CryptographicHashContent` the possible values are enumerated even though it's a drop-down list in the processor configuration rather than a freeform text field.

**Comment by pcgrenier** [comment]
> I don't think the error has to explicitly list all the available options, but I think the docs should. The error can just reference the docs to let people know where the definitive list is.
> 
> The list of supported algorithms does not change frequently, and the docs can be updated in the event the list changes.
> 
> There are some examples of our current hash usage in [`HashService`](https://github.com/apache/nifi/blob/master/nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/security/util/crypto/HashService.java#L59) and [`HashAlgorithm`](https://github.com/apache/nifi/blob/master/nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/security/util/crypto/HashAlgorithm.java). I believe in the docs for `CryptographicHashContent` the possible values are enumerated even though it's a drop-down list in the processor configuration rather than a freeform text field.

Ok, I added the list dynamically to the error message, also to the docs.

File: nifi-commons/nifi-expression-language/src/test/java/org/apache/nifi/attribute/expression/language/TestQuery.java (hunk 1 above)
Comment (pcgrenier) [reply nifi-commons/nifi-expression-language/src/test/java/org/apache/nifi/attribute/expression/language/TestQuery.java]: done.

**Comment by pcgrenier** [comment]
Going to close and rebase to master

**Comment by pcgrenier** [comment]
Closing to re-open on master

**Pull request closed without merging**